powerauth-tpp-engine/src/main/java/io/getlime/security/powerauth/app/tppengine/controller/TppRegistryController.java

```diff
@@ -43,13 +48,66 @@ public TppRegistryController(TppService tppService) {
-        final TppAppDetailResponse response = tppService.fetchAppDetailByClientId(clientId);
+    public ObjectResponse<TppAppDetailResponse> fetchAppInfoFromClientId(
+            @RequestParam("clientId") String clientId,
+            @RequestParam(value = "tppLicense", required = false) String tppLicense) throws TppAppNotFoundException, TppNotFoundException {
+        if (tppLicense != null) {
+            final TppAppDetailResponse response = tppService.fetchAppDetailByClientId(clientId, tppLicense);
+            return new ObjectResponse<>(response);
+        } else {
+            final TppAppDetailResponse response = tppService.fetchAppDetailByClientId(clientId);
+            return new ObjectResponse<>(response);
+        }
+    }
+
+    @RequestMapping(value = "app/list", method = RequestMethod.GET)
+    public ObjectResponse<List<TppAppDetailResponse>> getAppList(@RequestParam("tppLicense") String tppLicense) throws TppNotFoundException {
+        final List<TppAppDetailResponse> response = tppService.fetchAppListByTppLicense(tppLicense);
         if (response != null) {
             return new ObjectResponse<>(response);
         } else {
-            throw new TppAppNotFoundException(clientId);
+            throw new TppNotFoundException("tpp.notFound", tppLicense);
         }
     }
 
+    @RequestMapping(value = "app", method = RequestMethod.POST)
+    public ObjectResponse<TppAppDetailResponse> createApp(@RequestBody ObjectRequest<CreateTppAppRequest> request) throws UnableToCreateAppException {
+        final CreateTppAppRequest requestObject = request.getRequestObject();
+        final List<String> errors = CreateTppAppRequestValidator.validate(requestObject);
+        if (errors != null) { // request was not valid
+            throw new UnableToCreateAppException(errors);
+        }
+        final TppAppDetailResponse tppAppDetailResponse = tppService.createApp(requestObject);
+        return new ObjectResponse<>(tppAppDetailResponse);
+    }
+
+    @RequestMapping(value = "app", method = RequestMethod.PUT)
+    public ObjectResponse<TppAppDetailResponse> updateApp(
+            @RequestParam("clientId") String clientId,
+            @RequestBody ObjectRequest<CreateTppAppRequest> request) throws UnableToCreateAppException, TppNotFoundException, TppAppNotFoundException {
+        final CreateTppAppRequest requestObject = request.getRequestObject();
+        final List<String> errors = CreateTppAppRequestValidator.validate(requestObject);
+        if (errors != null) { // request was not valid
+            throw new UnableToCreateAppException(errors);
+        }
+        final TppAppDetailResponse tppAppDetailResponse = tppService.updateApp(clientId, requestObject);
+        return new ObjectResponse<>(tppAppDetailResponse);
+    }
+
+    @RequestMapping(value = "app/renewSecret", method = RequestMethod.POST)
```

Comment:
I am not sure about capitalized `S`. Perhaps this should be `app/secret/renew` or `app/renew-secret`?

Reply:
As defined in COBS standard.